File: ruby/tictactoe/lib/tictactoe/models/board.rb
module Tictactoe
  module Models
    class Board
      attr_accessor :size, :cells, :number_of_moves

      def initialize(size)
        self.size = size
        self.cells = Array.new(size) { Array.new(size) { nil } }
        self.number_of_moves = 0
      end

      def place(symbol, x, y)
        raise Tictactoe::Errors::InvalidCoordinatesError if x >= size || y >= size
        raise Tictactoe::Errors::BoardCellOccupiedError if cells[x][y]

        cells[x][y] = symbol
        self.number_of_moves += 1
      end

      def print
        line = ""
        (size * size).times { line += "-" }
        puts line

        0.upto(size - 1) do |row|
          row_line  = ""
          0.upto(size - 1) do |col|
            row_line += "#{cells[row][col]}  "
          end
          puts row_line
        end

        puts line
      end


      def has_winner?
        # no point checking winner case if atleast half board is empty
        return false if ( ((size * 2)  - 1 ) > number_of_moves )

        horizontal_winner? || vertical_winner? || diagonal_winner?
      end

      def draw?
        size * size == number_of_moves
      end

      private def horizontal_winner?
        0.upto(size - 1).each do |row|
          return true if cells[row].all? { |el| el == "O" } || cells[row].all? {|el| el == "X" }
        end

        false
      end

      private def vertical_winner?
        (0...size - 1).each do |col|
          return true if (0...size - 1).all? { |row| cells[row][col] == "O" } || (0...size).all? { |row| cells[row][col] == "X" }
        end

        false
      end

      private def diagonal_winner?
        straight_diagonal_winner? || opposite_diagonal_winner?
      end

      private def straight_diagonal_winner?
        first_element = cells[0][0]
        0.upto(size - 1).each do |diagonal|
          return false if cells[diagonal][diagonal] != first_element
        end

        true
      end

      private def opposite_diagonal_winner?
        first_element = cells[0][size - 1]
        (size - 1).downto(0).each do |diagonal|
          return false if cells[diagonal][size - diagonal - 1] != first_element
        end

        true
      end
    end
  end
end

File: java/tictactoe/src/main/java/com/gagan93jtg/lld_learning/models/Board.java
package com.gagan93jtg.lld_learning.models;

import com.gagan93jtg.lld_learning.errors.BoardCellOccupiedError;
import com.gagan93jtg.lld_learning.errors.InvalidCoordinatesError;

import java.util.Arrays;

public class Board {
  int size;
  String[][] cells;

  int numberOfMoves;

  public Board(int size) {
    this.size = size;
    cells = new String[size][size];
    this.numberOfMoves = 0;
  }

  public void place(String symbol, int x, int y) throws InvalidCoordinatesError, BoardCellOccupiedError {
    if (x >= size || y >= size) {
      throw new InvalidCoordinatesError("X and Y should be within bounds of board size: " + size);
    }

    if (cells[x][y] != null) {
      throw new BoardCellOccupiedError("Board cell at [" + x + "," + y + "] is already occupied");
    }

    cells[x][y] = symbol;
    this.numberOfMoves++;
  }

  public void print() {
    String line = "";
    for (int i = 0; i < size * size; i++) {
      line += "-";
    }
    System.out.println(line);
    for (int i = 0; i < size; i++) {
      String rowLine = "";
      for (int j = 0; j < size; j++) {
        rowLine += (cells[i][j] + " ");
      }
      System.out.print(rowLine);
    }
    System.out.println(line);
  }

  public boolean hasWinner() {
    // no point checking winner case if atleast half board is empty
    if (((size * 2) - 1) > numberOfMoves) {
      return false;
    }

    return hasHorizontalWinner() || hasVerticalWinner() || hasDiagonalWinner();
  }

  public boolean hasDraw() {
    return size * size == numberOfMoves;
  }

  private boolean hasHorizontalWinner() {
    for (int i = 0; i < size; i++) {
      if (Arrays.stream(cells[i]).allMatch(el -> el != null && el.equals("O")) || Arrays.stream(cells[i]).allMatch(el -> el != null &&  el.equals("X")))
        return true;
    }

    return false;
  }

  private boolean hasVerticalWinner() {
    for (int col = 0; col < size; col++) {
      boolean allOs = true;
      boolean allXs = true;
      for (int row = 0; row < size; row++) {
        if (cells[row][col] != "O") {
          allOs = false;
        }
        if (cells[row][col] != "X") {
          allXs = false;
        }
      }
      if (allOs || allXs) {
        return true;
      }
    }
    return false;
  }

  private boolean hasDiagonalWinner() {
    return hasStraightDiagonalWinner() || hasOppositeDiagonalWinner();
  }

  private boolean hasStraightDiagonalWinner() {
    String firstElement = cells[0][0];
    for (int i = 0; i < size; i++) {
      if (cells[i][i] != firstElement)
        return false;
    }
    return true;
  }

  private boolean hasOppositeDiagonalWinner() {
    String firstElement = cells[0][size - 1];
    for (int i = size - 1; i > 0; i--) {
      if (cells[i][size - i - 1] != firstElement)
        return false;
    }

    return true;
  }
}
